### PR TITLE
nodeset-check: check if used nodesets exist

### PR DESCRIPTION
**Pull request by pmpbaptista**
Warning when used nodesets don't exist in the project scope

Issue: #22

### CODE REVIEW

**Review by martimmatias — APPROVED** [review]

**Review by martimmatias — CHANGES_REQUESTED** [review]
```
Traceback (most recent call last):
  File "~/repos/github/zuulcilint/venv/bin/zuulcilint", line 6, in <module>
    sys.exit(main())
             ^^^^^^
  File "~/repos/github/zuulcilint/zuulcilint/__main__.py", line 342, in main
    inexistent_nodesets = zuul_checker.check_inexistent_nodesets(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/repos/github/zuulcilint/zuulcilint/checker.py", line 97, in check_inexistent_nodesets
    job_nodeset_list = job["job"]["nodeset"]["nodes"]
                       ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
TypeError: string indices must be integers, not 'str'
```

File: zuulcilint/checker.py [hunk 1]
```diff
@@ -49,12 +47,58 @@ def check_duplicated_jobs(
...
+            if isinstance(node["name"], str):
+                nodeset_list.add(node["name"])
+            if isinstance(node["name"], list):
+                for node_name in node["name"]:
+                    nodeset_list.add(node_name)
+    inexistent_nodesets = set()
+
+    for job in jobs:
+        try:
+            job_nodeset_list = job["job"]["nodeset"]["nodes"]
```
Comment (martimmatias) [inline zuulcilint/checker.py]: The nodeset property of Job can be a string, not only an object.

**Review by martimmatias — CHANGES_REQUESTED** [review]

**Review by martimmatias — DISMISSED** [review]

File: zuulcilint/__main__.py [hunk 2]
```diff
@@ -93,70 +132,113 @@ def get_all_zuul_yaml_files(files: list[str]) -> list[pathlib.Path]:
     return zuul_yaml_files
 
 
-def get_all_jobs(zuul_yaml_files: list[pathlib.Path]) -> list[list[str]]:
-    """Get all jobs from Zuul YAML files."""
-    all_jobs = []
+def get_all_zuul_objects_by_type(
+    zuul_yaml_files: List[pathlib.Path],
+    zuul_obj: ZuulObject,
+) -> List[Union[dict, None]]:
```
Comment (martimmatias) [inline zuulcilint/__main__.py]: Either use "Union" or the pipe "|" operator which is already used in the DocString for zuul_utils.get_zuul_object_from_yaml.

File: zuulcilint/__main__.py [hunk 3]
```diff
@@ -93,70 +132,113 @@ def get_all_zuul_yaml_files(files: list[str]) -> list[pathlib.Path]:
...
+    Returns:
+    -------
+        A list of Zuul objects.
+    """
+    all_zuul_objects = []
     for file_path in zuul_yaml_files:
-        jobs = zuul_utils.get_zuul_object_from_yaml(ZuulObject.JOB, file_path)
-        all_jobs.append([job.get(ZuulObject.JOB.value, {}).get("name") for job in jobs])
-    return all_jobs
+        zuul_objects = zuul_utils.get_zuul_object_from_yaml(zuul_obj, file_path)
```
Comment (martimmatias) [inline zuulcilint/__main__.py:153]: In what cases does zuul_utils.get_zuul_object_from_yaml return None? Doesn't it always return a list (even if empty)?

File: zuulcilint/__main__.py [hunk 4]
```diff
@@ -93,70 +132,113 @@ def get_all_zuul_yaml_files(files: list[str]) -> list[pathlib.Path]:
...
         zuul_utils.print_bold(
-            f"Found {n_bad} files with 'yml' extension",
+            f"Found {n_bad_yaml} files with 'yml' extension",
             None,
         )
-        for file_path in bad_yml_files:
+        for file_path in warnings["warnings"]["bad_yaml_files"]:
             print(f"{file_path}")
-    if duplicated_jobs:
+    if n_duplicate:
```
Comment (martimmatias) [inline zuulcilint/__main__.py:193]: Add a newline before this if statement for readability.

**Review by martimmatias — CHANGES_REQUESTED** [review]

File: zuulcilint/__main__.py (hunk 2 above)
Comment (pmpbaptista) [reply zuulcilint/__main__.py]: I've added the annotations import which fixes the syntax so we can use pipe and the known types (dict,list...)

File: zuulcilint/__main__.py (hunk 3 above)
Comment (pmpbaptista) [reply zuulcilint/__main__.py:153]: Removed None

File: zuulcilint/__main__.py (hunk 4 above)
Comment (pmpbaptista) [reply zuulcilint/__main__.py:193]: newline added

File: zuulcilint/__main__.py (hunk 3 above)
Comment (martimmatias) [reply zuulcilint/__main__.py:153]: Why did you type it as just "list" and not a list of dicts?

Comment (pmpbaptista) [reply zuulcilint/__main__.py:153]: Deleted while adjusting types. It's back now

File: zuulcilint/checker.py [hunk 5]
```diff
@@ -6,40 +6,49 @@
...
 def check_job_playbook_paths(
-    job: dict[str, str | list[str] | None],
+    job: list[dict[str, str | list[str]] | None],
 ) -> list[str]:
     """Check that all playbooks in a job have a valid path.
 
     Args:
     ----
-        job: A dictionary representing a Zuul job.
+        job: A list of Zuul jobs.
```
Comment (martimmatias) [inline zuulcilint/checker.py]: Job is not a list. Job is a dictionary of str to (at least) str | dict | list[str, dict].

File: zuulcilint/checker.py [hunk 6]
```diff
@@ -6,40 +6,49 @@
...
 def check_duplicated_jobs(
-    jobs: list[list[dict[str, str] | None]],
+    jobs: dict[pathlib.Path, list[dict | None]]
 ) -> set[dict[str, str] | None]:
     """Check that all jobs are unique in different Zuul YAML files.
 
     Args:
     ----
-        jobs: A list of lists of dictionaries representing Zuul jobs.
+        jobs: A list of Zuul jobs.
```
Comment (martimmatias) [inline zuulcilint/checker.py]: "jobs" is not a list

File: zuulcilint/utils.py [hunk 7]
```diff
@@ -124,6 +124,9 @@ def get_zuul_object_from_yaml(
     except yaml.YAMLError:
         print(f"Invalid YAML in Zuul YAML file: {zuul_yaml_file}", file=sys.stderr)
         sys.exit(1)
+    except AttributeError:
```
Comment (martimmatias) [inline zuulcilint/utils.py:127]: When is this exception raised?

**Review by martimmatias — CHANGES_REQUESTED** [review]

File: zuulcilint/utils.py (hunk 7 above)
Comment (pmpbaptista) [reply zuulcilint/utils.py:127]: When trying to lint paths that don't contain zuul config files .. this way we exit with a clean error

Comment (pmpbaptista) [reply zuulcilint/utils.py:127]: When trying to lint paths that don't contain zuul config files .. this way we exit with a clean error

Comment (martimmatias) [reply zuulcilint/utils.py:127]: Why that exception specifically?

Comment (pmpbaptista) [reply zuulcilint/utils.py:127]: > Checking playbook paths
> Traceback (most recent call last):
>   File "/home/pbaptista/Workspace/private/zuulcilint/venv/bin/zuulcilint", line 6, in <module>
>     sys.exit(main())
>   File "/home/pbaptista/Workspace/private/zuulcilint/zuulcilint/__main__.py", line 322, in main
>     invalid_playbook_paths = lint_playbook_paths(zuul_good_yaml)
>   File "/home/pbaptista/Workspace/private/zuulcilint/zuulcilint/__main__.py", line 109, in lint_playbook_paths
>     jobs = zuul_utils.get_zuul_object_from_yaml(ZuulObject.JOB, file_path)
>   File "/home/pbaptista/Workspace/private/zuulcilint/zuulcilint/utils.py", line 120, in get_zuul_object_from_yaml
>     return [obj for obj in yaml.safe_load(f) if obj.get(obj_type.value)]
>   File "/home/pbaptista/Workspace/private/zuulcilint/zuulcilint/utils.py", line 120, in <listcomp>
>     return [obj for obj in yaml.safe_load(f) if obj.get(obj_type.value)]
> AttributeError: 'str' object has no attribute 'get'

**Review by martimmatias — APPROVED** [review]

**Pull request merged**